dll/win32/shell32/CDefView.cpp

```diff
@@ -3918,16 +3929,35 @@ HRESULT STDMETHODCALLTYPE CDefView::CreateViewWindow3(IShellBrowser *psb, IShell
     return S_OK;
 }
 
-HRESULT STDMETHODCALLTYPE CDefView::HandleRename(LPCITEMIDLIST new_pidl)
+HRESULT STDMETHODCALLTYPE CDefView::HandleRename(LPCITEMIDLIST pidl)
 {
-    FIXME("(%p)->(%p) stub\n", this, new_pidl);
-    return E_NOTIMPL;
+    if (!pidl)
+    {
+        int idx = m_ListView.GetNextItem(-1, LVNI_SELECTED | LVNI_FOCUSED);
+        if (idx < 0)
+            idx = m_ListView.GetNextItem(-1, LVNI_SELECTED);
+        pidl = _PidlByItem(idx);
+    }
+    if (ILFindLastID(pidl) != pidl)
+        return E_INVALIDARG;
+    return SelectItem(pidl, SVSI_EDIT | SVSI_ENSUREVISIBLE | SVSI_FOCUSED | SVSI_SELECT);
+}
+
+HRESULT CDefView::SelectAndPositionItem(int Idx, UINT fSVSI, POINT *ppt)
+{
+    if (Idx == -1)
+        return fSVSI == SVSI_DESELECTOTHERS ? SelectItem(-2, fSVSI) : E_INVALIDARG;
+    if (ppt)
+        m_ListView.SetItemPosition(Idx, ppt);
+    return SelectItem(Idx, fSVSI); // After SetItemPosition for SVSI_ENSUREVISIBLE
 }
 
 HRESULT STDMETHODCALLTYPE CDefView::SelectAndPositionItem(LPCITEMIDLIST item, UINT flags, POINT *point)
 {
-    FIXME("(%p)->(%p, %u, %p) stub\n", this, item, flags, point);
-    return E_NOTIMPL;
+    if (!item)
+        return SelectAndPositionItem(-1, flags, point);
+    int idx = LV_FindItemByPidl(item);
+    return idx != -1 ? SelectAndPositionItem(idx, flags, point) : S_FALSE;
```

Comment:
Alt version where `SelectAndPositionItem()` is invoked only in one place:
```suggestion
    int idx = -1;
    if (item && (idx = LV_FindItemByPidl(item)) == -1)
        return S_FALSE;
    return SelectAndPositionItem(idx, flags, point);
```